tofhir-server/src/main/scala/io/tofhir/server/service/mappingcontext/MappingContextFolderRepository.scala

```diff
@@ -0,0 +1,179 @@
+import io.tofhir.engine.model.FhirMappingException
+import io.tofhir.engine.util.FileUtils
+import io.tofhir.server.model._
+import io.tofhir.server.service.project.ProjectFolderRepository
+import io.tofhir.server.util.FileOperations
+
+import java.io.File
+import scala.collection.mutable
+import scala.concurrent.Future
+
+/**
+ * Folder/Directory based mapping context repository implementation.
+ *
+ * @param mappingContextRepositoryFolderPath root folder path to the mapping context repository
+ */
+class MappingContextFolderRepository(mappingContextRepositoryFolderPath: String, projectFolderRepository: ProjectFolderRepository) extends IMappingContextRepository {
+  // project id -> mapping context id
+  private val mappingContextDefinitions: mutable.Map[String, Seq[String]] = initMap(mappingContextRepositoryFolderPath)
+
+  /**
+   * Returns the mapping context cached in memory
+   *
+   * @return
+   */
+  def getCachedMappingContexts(): mutable.Map[String, Seq[String]] = {
+    mappingContextDefinitions
+  }
+
+  /**
+   * Retrieve the metadata of all mapping context ids
+   *
+   * @return
+   */
+  override def getAllMappingContext(projectId: String): Future[Seq[String]] = {
+    Future {
+      if (mappingContextDefinitions.contains(projectId)) {
+        mappingContextDefinitions(projectId)
+      } else {
+        Seq.empty
+      }
+    }
+  }
+
+  /**
+   * Save the mapping context to the repository.
+   * We only store id of a mapping context in the project metadata json file.
+   *
+   * @param projectId project id the mapping context belongs to
+   * @param id        mapping context id to save
+   * @return
+   */
+  override def createMappingContext(projectId: String, id: String): Future[String] = {
+    Future {
+      if (mappingContextDefinitions.contains(projectId) && mappingContextDefinitions(projectId).contains(id)) {
+        throw AlreadyExists("Fhir mapping already exists.", s"A mapping context definition with id ${id} already exists in the mapping context repository at ${FileUtils.getPath(mappingContextRepositoryFolderPath).toAbsolutePath.toString}")
+      }
+      // Write to the repository as a new file
+      getFileForMappingContext(projectId, id).map(newFile => {
+        newFile.createNewFile()
+      })
+      // Add to the project metadata json file
+      projectFolderRepository.addMappingContext(projectId, id)
+      // Add to the in-memory map
+      mappingContextDefinitions(projectId) = mappingContextDefinitions.getOrElseUpdate(projectId, Seq.empty) :+ id
+      id
+    }
+  }
+
+
+  /**
+   * Delete the mapping context from the repository
+   *
+   * @param projectId project id the mapping context belongs to
+   * @param id        mapping context id
+   * @return
+   */
+  override def deleteMappingContext(projectId: String, id: String): Future[Unit] = {
+    Future {
+      if (!mappingContextDefinitions.contains(projectId) || !mappingContextDefinitions(projectId).contains(id)) {
```

Comment:
It is repeated three times. We can create a method to check the existence of a mapping context and use it.